modules/partition-replicator/src/main/java/org/apache/ignite/internal/partition/replicator/PartitionReplicaLifecycleManager.java

```diff
@@ -1264,4 +1302,52 @@ private void cleanUpPartitionsResources(Set<ReplicationGroupId> partitionIds) {
             LOG.error("Unable to clean zones resources", e);
         }
     }
+
+    /**
+     * Lock the zones replica list for any changes. {@link #hasLocalPartition(ZonePartitionId)} must be executed under this lock always.
+     *
+     * @param zoneId Zone id.
+     * @return Stamp, which must be used for further unlock.
+     */
+    public long lockZoneForRead(int zoneId) {
+        AtomicLong stamp = new AtomicLong();
+
+        zonePartitionsLocks.compute(zoneId, (id, l) -> {
+            if (l == null) {
+                l = new StampedLock();
+            }
+
+            stamp.set(l.readLock());
+
+            return l;
+        });
+
+        return stamp.get();
+    }
+
+    /**
+     * Unlock zones replica list.
+     *
+     * @param zoneId Zone id.
+     * @param stamp Stamp, produced by the according {@link #hasLocalPartition(ZonePartitionId) call.}
+     */
+    public void unlockZoneForRead(int zoneId, long stamp) {
+        zonePartitionsLocks.get(zoneId).unlockRead(stamp);
+    }
+
+    /**
+     * Load a new table partition listener to the zone replica.
+     *
+     * @param zonePartitionId Zone partition id.
+     * @param tablePartitionId Table partition id.
+     * @param createListener Lazy replica listener from RAFT command runner builder.
+     */
+    public void loadTableListenerToZoneReplica(ZonePartitionId zonePartitionId, TablePartitionId tablePartitionId,
+            Function<RaftCommandRunner, ReplicaListener> createListener) {
+        CompletableFuture<Replica> replicaFut = replicaMgr.replica(zonePartitionId);
+
+        assert replicaFut != null && replicaFut.isDone();
+
+        ((ZonePartitionReplicaListener) replicaFut.join().listener()).addTableReplicaListener(tablePartitionId, createListener);
```

Comment:
Because of previous assert - I think that it will be better and cleaner. Moreover we use this pattern in our code here and there.

But I added javadoc for the method, that we need to have such guarantee.